### PR TITLE
Expanding documentation

### PR DESCRIPTION
**Pull request by liz-is**
I've updated the docs to include:

- background info from https://www.hanachockler.com/rex/
- usage info from README (README now less info: quickstart, example results, and links to docs)
- a page with all possible parameters that can be set in `rex.toml` -- @kellino this will need your input for a few cases where things were unclear, see the comments in docs/config.md!
- the logo

Feedback welcome!

A couple of specific questions I had:

- there's an `assets/` and an `imgs/` folder, but it's not super clear to me what goes in which - could these be merged?
- I think `scripts/example_preprocess.py` is missing or has been renamed...
- is there a larger version of the ReX logo? I think it would look nice to have a larger version on the docs pages

Edit to add the link to the rendered updated docs: https://rex-xai--7.org.readthedocs.build/en/7/

### CODE REVIEW

**Comment by kellino** [comment]
> I've updated the docs to include:
> 
>     * background info from https://www.hanachockler.com/rex/
> 
>     * usage info from README (README now less info: quickstart, example results, and links to docs)
> 
>     * a page with all possible parameters that can be set in `rex.toml` -- @kellino this will need your input for a few cases where things were unclear, see the comments in docs/config.md!
> 
>     * the logo
> 
> 
> Feedback welcome!
> 
> A couple of specific questions I had:
> 
>     * there's an `assets/` and an `imgs/` folder, but it's not super clear to me what goes in which - could these be merged?
> 

`imgs/` probably shouldn't be in the repo at all, thinking clearly about it. I just use these as convenient standard test images. It certainly shouldn't be merged into main anyway, but it's quite useful to have a few images attached to the repo if I'm on a different computer. `assets/` should only contain files associated with the README, but merging would not be a disaster. 

>     * I think `scripts/example_preprocess.py` is missing or has been renamed...

yeah, I think it doesn't exist anymore. It's probably what became `scripts/pytorch.py`. That interface has changed a lot
> 
>     * is there a larger version of the ReX logo? I think it would look nice to have a larger version on the docs pages
> 

I can make one. I've got the original artwork somewhere as a svg file, so it should be easy to scale it up
> 
> Edit to add the link to the rendered updated docs: https://rex-xai--7.org.readthedocs.build/en/7/

**Comment by liz-is** [comment]
> I just use these as convenient standard test images. 

I think we should definitely keep some test images in the repo! I've been using them to do informal testing of changes, and we can in due course also use them for formal automated testing. But maybe we can move these test images inside of `tests/`, and keep `assets/` for images used in the README/docs.

> [`scripts/preprocess.py`] doesn't exist anymore.

Okay, no problem, I'll rephrase this bit

> I can make one. I've got the original artwork somewhere as a svg file, so it should be easy to scale it up

That would be great! There is an svg file in the repo, but it seems like it comtains two different variants of the logo, so I didn't use it as-is.

File: docs/index.rst [hunk 1]
```diff
@@ -3,15 +3,91 @@
...
+We assume that we can modify the inputs and send them to the classifier, observing the output.
+ReX outperforms other tools on single explanations, non-contiguous explanations (for partially obscured images), and multiple explanations.
 
+.. image:: ../assets/rex-structure-600x129.png
+  :width: 600
+  :alt: ReX organisation
+
+For more information and links to the papers, see the :doc:`background` page.
+
+Installation
```
Comment (LouiseABowler) [inline docs/index.rst:20]: The installation and quickstart sections look to be the same as in the README, apart from `.md`/`.rst` formatting. Could you import them from one file into the other? Maybe something like [this](https://muffinresearch.co.uk/selectively-including-parts-readme-rst-in-your-docs/) could work?

File: docs/index.rst [hunk 2]
```diff
@@ -3,15 +3,91 @@
...
+   git clone git@github.com:ReX-XAI/ReX.git
+   cd ReX/
+
+We recommend creating a virtual environment to install ReX.
+ReX has been tested using versions of Python >= 3.10.
+The following instructions assume ``conda``::
+
+   conda create -n rex python=3.12
+   conda activate rex
+   pip install .
```
Comment (LouiseABowler) [inline docs/index.rst:34]: Should this be switched to use poetry now?

File: README.md [hunk 3]
```diff
@@ -8,99 +8,94 @@
 
 <!--- BADGES: START --->
 
-[![CI Pipeline](https://github.com/kellino/ReX/actions/workflows/ci-pipeline.yml/badge.svg)](https://github.com/kellino/ReX/actions/workflows/ci-pipeline.yml)
+[![CI Pipeline](https://github.com/ReX-XAI/ReX/actions/workflows/python-package.yml/badge.svg)](https://github.com/ReX-XAI/ReX/actions/workflows/python-package.yml)
 [![License](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/kellino/ReX.jl/blob/main/LICENSE)
```
Comment (LouiseABowler) [inline README.md:12]: The license link here needs an update after the repo move

**Review by LouiseABowler — COMMENTED** [review]
The docs pages are looking really good! I've left some suggestions, let me know when you're ready for me to have another look and send a ✅ over @liz-is.

File: docs/index.rst (hunk 2 above)
Comment (liz-is) [reply docs/index.rst:34]: I don't think it _has_ to be - poetry installation instructions are in the CONTRIBUTING.md but a regular user wouldn't need to use poetry, they can install with whatever their usual workflow is. 

My personal feeling is that given the difficulties poetry's upper version pinning can cause, I'd rather not encourage regualr users to use it 😅 (even if we're using it as a build tool!)

File: docs/background.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# Background information
+
+ReX is a causal explainability tool for image classifiers.
+ReX is black-box, that is, agnostic to the internal structure of the classifier.
+We assume that we can modify the inputs and send them to the classifier, observing the output.
+ReX outperforms other tools on single explanations, non-contiguous explanations (for partially obscured images), and multiple explanations.
```
[the remark this answers is not in the document]
Comment (liz-is) [reply docs/background.md:6]: Done - added links to the three different papers which cover the different claims here.

File: docs/background.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+![ReX organisation](../assets/rex-structure-600x129.png)
+
+## Assumptions
```
[the remark this answers is not in the document]
Comment (liz-is) [reply docs/background.md:10]: That's fair, it's not super clear. I'll add some additional explanatory text. My understanding is that this aims to convey something like "ReX works on the assumption that if we can intervene on the inputs to a model and observe changes in its outputs, we can use this information to reason about the way the DNN makes its decisions".

File: docs/command_line.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+Some options from the config file can be overridden at the command line when calling ReX.
+In particular, you can change the number of iterations of the algorithm:
+
+```bash
+rex <path_to_image> --model <path_to_model>  --iters 5
+```
+
+## Preprocessing
+
+ReX by default tries to make reasonable guesses for image preprocessing.
```
[the remark this answers is not in the document]
Comment (liz-is) [reply docs/command_line.md:96]: The preprocessing has a few different aspects. My understanding for images:

- it guesses the mode of the input image (e.g. greyscale or RGB)
- if necessary converts greyscale input to pseudo-RGB if the model requires this
  - I just realised that this is the _opposite_ of what is actually written here: "In the event the the model input is single channel and the image is multi-channel, then ReX will try to convert the image to greyscale."
  - @kellino am I misunderstanding what the code does, or could you clarify? Should the code be able to do both conversions? 
- resizes the input to match the model size
- converts to a numpy array of float32
- possibly transposes the order of the data axes (height, width, colour) to match the order required by the model
- creates a Tensor from the numpy array
- normalises the data by dividing by 255 (NB this is always done, `args.norm` is not used)
- then optionally does further normalisation, depending on the values of `args.means` and `args.stds`

For .npy files with `args.mode` tabular or spectral:

- it loads the data with `np.load`
- converts the array to float32
- doesn't resize but does add additional axes if the data has fewer axes than the model shape
- converts the numpy array to a Tensor

So overall my understanding is that the prepreocessing will always involve, for images at least:
- resizing the input and adjusting the order of the axes to match the expectations of the model
- normalising by dividing by 255
- converting to a Tensor
and optionally:
- normalisation using `args.means` and `args.stds`

So I'll think about how to rephrase to include these steps more clearly!

Although, I also realised that right now the `--processed` flag doesn't actually do anything, as far as I can tell, so maybe the example should be removed for now...

File: docs/index.rst (hunk 1 above)
Comment (liz-is) [reply docs/index.rst:20]: Nice find - I knew it was possible to imclude a file, I didn't know you could include only part of a file! I think I got this working now.

**Comment by liz-is** [comment]
I think all the `docs/config.md` updates we discussed have now been implemented!

We also discussed adding separate pages to clarify some aspects of the core logic of ReX (and probably add diagrams). The key pages/sections I think we'd need are:

- Mutant generation, and how the tree of mutants works, to help explain `tree_depth`, `search_limit`, etc
- Something to explain how `queue_style` works
- How the global method for finding an explanation works - explaining `chunk_size`
- How the spatial method for finding an explanation works - explaining `initial_radius`, `radius_eta`, etc
- (Later) How multi-explanations work, including the `obj_function` option.

Separately, it'd be nice (but not high priority IMO) to have a page/section about the evaluation metrics with e.g. links to papers defining/using these.

However, I think the points above are not *essential* for this version of the docs, and could be added later. What do you think (@LouiseABowler and @kellino)?

A couple of things I do think we should clarify for this version (@kellino requesting your input here!):
- The docs say "In the event the the model input is single channel and the image is multi-channel, then ReX will try to convert the image to greyscale." - but my understanding from the code is that it does the opposite of this, i.e. converts a greyscale image to pseudo-RGB if the model requires RGB input. Can you confirm what the intended behaviour is?
- It looks like the `--processed` flag is not currently used in the code. Should I remove this from the docs, or should I implement its usage?

**Comment by LouiseABowler** [comment]
Filling in those new pages later on sounds good to me. Do you want to open a new issue with the list of pages in there, and then we've got a clear reminder about them?

File: docs/index.rst (hunk 2 above)
Comment (LouiseABowler) [reply docs/index.rst]: Agree with you that it doesn't have to be installed with poetry. But it feels unusual to be mixing and matching tools without a reason - if you want to keep it as is, could there be some docs somewhere explaining why? Maybe in the contributing file if that's where the poetry method is introduced?

Comment (liz-is) [reply docs/index.rst]: That's fair. I've added a brief explanation about why we use poetry in the contributing file!

**Pull request merged**